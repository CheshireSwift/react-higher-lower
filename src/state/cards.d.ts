export type Suit = "H" | "S" | "C" | "D";
export type Rank = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13;
export interface Card {
  rank: Rank;
  suit: Suit;
}
