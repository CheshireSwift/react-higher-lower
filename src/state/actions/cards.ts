import _ from "lodash";

import { Card, Guess } from "../cards";

export const draw = (card: Card) => ({
  type: "DRAW" as "DRAW",
  payload: { card }
});

export const guess = (direction: Guess, card: Card) => ({
  type: "GUESS" as "GUESS",
  payload: { card, direction }
});

export const guessHigher = _.partial(guess, "HIGHER");
export const guessLower = _.partial(guess, "LOWER");

export const shuffle = () => ({
  type: "SHUFFLE" as "SHUFFLE"
});
