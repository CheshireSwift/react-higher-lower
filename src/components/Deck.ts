import React from "react";

import { div, button, h } from "react-hyperscript-helpers";
import { Dispatch, bindActionCreators } from "redux";
import { connect } from "react-redux";

import { State } from "../state";
import { Card } from "../state/cards";
import { shuffle, draw, guessHigher, guessLower } from "../state/actions/cards";
import { CardDisplay } from "./CardDisplay";

interface DeckProps {
  cards: Card[];
  draw: (card: Card) => any;
  guessHigher: (card: Card) => any;
  guessLower: (card: Card) => any;
  shuffle: () => any;
}

export class Deck extends React.PureComponent<DeckProps> {
  componentDidMount() {
    this.props.shuffle();
  }

  render() {
    return div({ style: { position: "relative" } }, [
      this.props.cards.length,
      h(CardDisplay, { card: null }),
      button({ onClick: () => this.props.draw(this.props.cards[0]) }, "draw"),
      button(
        { onClick: () => this.props.guessHigher(this.props.cards[0]) },
        "higher"
      ),
      button(
        { onClick: () => this.props.guessLower(this.props.cards[0]) },
        "lower"
      )
    ]);
  }
}

export default connect(
  (state: State) => ({ cards: state.deck }),
  (dispatch: Dispatch) =>
    bindActionCreators({ draw, shuffle, guessHigher, guessLower }, dispatch)
)(Deck);
