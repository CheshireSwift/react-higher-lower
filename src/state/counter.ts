import { increment, decrement, set } from "./actions/tick";

interface CounterState {
  count: number;
}

export const counter: Reducer<
  CounterState,
  typeof increment | typeof decrement | typeof set
> = (state = { count: 0 }, action) => {
  switch (action.type) {
    case "INCREMENT":
      return { ...state, count: state.count + action.payload.amount };
    case "DECREMENT":
      return { ...state, count: state.count - action.payload.amount };
    case "SET":
      return { ...state, count: action.payload.value };
    default:
      return state;
  }
};
