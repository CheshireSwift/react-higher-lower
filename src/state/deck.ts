import _ from "lodash";

import { Card, Rank, Suit } from "./cards";
import { draw, shuffle } from "./actions/cards";

type DeckState = Card[];

const sortedDeck = (): DeckState =>
  _.flatten(
    (["S", "H", "C", "D"] as Suit[]).map(suit =>
      ([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13] as Rank[]).map(rank => ({
        suit,
        rank
      }))
    )
  );

export const deck: Reducer<DeckState, typeof draw | typeof shuffle> = (
  state = sortedDeck(),
  action
) => {
  switch (action.type) {
    case "DRAW":
      return _.reject(state, action.payload);
    case "SHUFFLE":
      return _.shuffle(state);
    default:
      return state;
  }
};
