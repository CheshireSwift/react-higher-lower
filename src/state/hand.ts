import _ from "lodash";

import { Card, Guess } from "./cards";
import { draw, guess } from "./actions/cards";

type HandState = Array<{ card: Card; guess: Guess | null }>;

export const hand: Reducer<HandState, typeof draw | typeof guess> = (
  state = [],
  action
) => {
  switch (action.type) {
    case "DRAW":
    case "GUESS":
      return [
        ...state,
        {
          card: action.payload.card,
          guess: _.get(action.payload, "direction", null)
        }
      ];
    default:
      return state;
  }
};
