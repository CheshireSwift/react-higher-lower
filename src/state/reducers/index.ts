import { combineReducers, Reducer } from "redux";

import { counter } from "./counter";

export type ReducerState<R> = R extends Reducer<infer S> ? S : null;
export const rootReducer = combineReducers({ counter });
