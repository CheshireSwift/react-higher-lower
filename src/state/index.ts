import { ReducerState, rootReducer } from "./reducers";
export type State = ReducerState<typeof rootReducer>;

export default State;
