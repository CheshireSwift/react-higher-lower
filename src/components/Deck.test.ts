import { render, shallow, ShallowWrapper } from "enzyme";
import { h } from "react-hyperscript-helpers";

import { Deck } from "./Deck";
import { Card } from "../state/cards";

describe("the Deck", () => {
  const cards: Card[] = [
    { rank: 1, suit: "S" },
    { rank: 3, suit: "C" },
    { rank: 7, suit: "D" }
  ];

  const props = {
    cards,
    draw: jest.fn(),
    guessHigher: jest.fn(),
    guessLower: jest.fn(),
    shuffle: jest.fn()
  };

  let deck: Cheerio;

  const findButton = (wrapper: ShallowWrapper<any>, text: string) =>
    wrapper.findWhere(
      child => child.type() === "button" && child.text() === text
    );

  beforeEach(() => {
    deck = render(h(Deck, props));
  });

  it("matches the snapshot", () => {
    expect(deck).toMatchSnapshot();
  });

  it("shows the remaining card count", () => {
    expect(deck.text()).toContain(cards.length);
  });

  it("draws a card when the button is clicked", () => {
    const drawButton = findButton(shallow(h(Deck, props)), "draw");
    const onClick: any = drawButton.prop("onClick");
    onClick();
    expect(props.draw).toHaveBeenCalled();
  });

  it("guesses high when the button is clicked", () => {
    const higherButton = findButton(shallow(h(Deck, props)), "higher");
    const onClick: any = higherButton.prop("onClick");
    onClick();
    expect(props.guessHigher).toHaveBeenCalled();
  });

  it("guesses low when the button is clicked", () => {
    const lowerButton = findButton(shallow(h(Deck, props)), "lower");
    const onClick: any = lowerButton.prop("onClick");
    onClick();
    expect(props.guessLower).toHaveBeenCalled();
  });

  it("shuffles the deck when mounted", () => {
    shallow(h(Deck, props));
    expect(props.shuffle).toHaveBeenCalled();
  });
});
