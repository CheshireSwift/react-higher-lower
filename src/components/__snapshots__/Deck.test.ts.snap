// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Deck matches the snapshot 1`] = `
<div
  style="position:relative"
>
  3
  <div
    style="width:40px;height:60px;border:1px solid black;background:radial-gradient(circle at 0% 50%, rgba(96, 16, 48, 0) 9px, #613 10px, rgba(96, 16, 48, 0) 11px) 0px 10px,
            radial-gradient(at 100% 100%,      rgba(96, 16, 48, 0) 9px, #613 10px, rgba(96, 16, 48, 0) 11px),
            #8a3;background-size:20px 20px"
  />
  <button>
    draw
  </button>
  <button>
    higher
  </button>
  <button>
    lower
  </button>
</div>
`;
